### DISCUSSION TITLE
introducing-ihostlifetime-and-untangling-the-generic-host-startup-interactions/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#23] [Announcements]
# introducing-ihostlifetime-and-untangling-the-generic-host-startup-interactions/

In this post I introduce the new IHostLifetime interface and look at the interactions involved in the ASP.NET Core generic host startup and shutdown processes

https://andrewlock.net/introducing-ihostlifetime-and-untangling-the-generic-host-startup-interactions/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 27 2019, 10:24, in reply to [David McClelland](https://disqus.com/by/disqus_8sWkGmnmug/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/23#discussioncomment-6057770)</em>

---

<p>Thanks! I've replied on StackOverflow 🙂</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 09 2019, 10:03, in reply to [Maciek Misztal](https://disqus.com/by/maciekmisztal/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/23#discussioncomment-6057775)</em>

---

<p>ServerFeatures is a web-specific notion, so it's available on IServer rather than IHost. You can get the IServer from IHost using</p><p><code></code></p><pre><code>IHost host;<br>var server = host.Services.GetRequiredService&lt;iserver&gt;();<br>server.Features.Get&lt;iserveraddressfeature&gt;()</code></pre><p></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 09 2019, 09:57, in reply to [Mark Davies](https://disqus.com/by/disqus_XkFi3rCYje/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/23#discussioncomment-6057780)</em>

---

<p>Have you heard of Oakton? <a href="https://jasperfx.github.io/oakton/documentation/aspnetcore/" rel="nofollow noopener" title="https://jasperfx.github.io/oakton/documentation/aspnetcore/">https://jasperfx.github.io/oakton/documentation/aspnetcore/</a></p><p>You can use it to easily build .NET CLI tools, and looks like it would solve many of the issues you're describing here 🙂</p>

[a comment by an automated account is not shown]
